cryptarchia/cryptarchia.py

```diff
@@ -46,8 +46,30 @@ class Config:
 class BlockHeader:
     slot: Slot
     parent: Id
-    # TODO: spec out the block id, this is just a placeholder to unblock tests
-    id: Id
+    content_size: int
+    content_id: Id
+
+    # **Attention**:
+    # The ID of a block header is defined as the 32byte blake2b hash of its fields
+    # as serialized in the format specified by the 'HEADER' rule in 'messages.abnf'.
+    #
+    # The following code is to be considered as a reference implementation, mostly to be used for testing.
+    def id(self) -> Id:
+        # version byte
+        bytes = bytearray(b"\x01")
+        # header type
+        bytes += b"\x00"
+        # content size
+        bytes += int.to_bytes(self.content_size, length=4, byteorder="big")
+        # content id
+        assert len(self.content_id) == 32
+        bytes += self.content_id
+        # slot
+        bytes += int.to_bytes(self.slot.absolute_slot, length=8, byteorder="big")
+        # parent
+        assert len(self.content_id) == 32
+        bytes += self.parent
+        return blake2b(bytes, digest_size=32).digest()
```

Comment:
I believe the correct way to do this sort of hashing is to feed bytes to the hasher object instead of concatenating all bytes together and then hashing. e.g.

```python3
h = blake2b(digest_size=32)
h.update(b"\x01") # version byte
h.update(b"\x00") # header type for block
# ...
return h.digest()
```

Reply:
curious about the difference

Reply:
they produce the same result, but using the update api is what you will see in practice everywhere. Main reason is avoiding unnecessary allocations while building up the bytes buffer